aptos-move/move-examples/nfts_as_accounts/sources/token.move

```diff
@@ -0,0 +1,232 @@
+        collection: String,
+        description: String,
+        mutability_config: MutabilityConfig,
+        name: String,
+        royalty: Royalty,
+        uri: String,
+        data: Data,
+    ): TokenRef<Data> {
+        let seed = *string::bytes(&collection);
+        vector::append(&mut seed, b"::");
+        vector::append(&mut seed, *string::bytes(&name));
+        // To keep costs down, this function does not check to see if the token already exists
+        let (token_account, inner) = account::create_resource_account(creator, seed);
+        let token = Token {
+            collection,
+            creator: signer::address_of(creator),
+            description,
+            mutability_config,
+            name,
+            royalty,
+            uri,
+            data: option::some(data),
+        };
+
+        move_to(&token_account, token);
+        TokenRef<Data> { inner }
+    }
+
+    public fun create_mutability_config(description: bool, name: bool, uri: bool): MutabilityConfig {
+        MutabilityConfig {
+            description,
+            name,
+            uri,
+        }
+    }
+
+    public fun create_royalty(numerator: u64, denominator: u64, payee_address: address): Royalty {
+        Royalty {
+            numerator,
+            denominator,
+            payee_address,
+        }
+    }
+
+    public fun generate_token_address(creator: address, collection: &String, name: &String): address {
+        let seed = *string::bytes(collection);
+        vector::append(&mut seed, b"::");
+        vector::append(&mut seed, *string::bytes(name));
+        account::create_resource_address(&creator, seed)
+    }
+
+    public fun to_base_token<Data>(ref: TokenRef<Data>): TokenRef<BaseToken> {
+        let TokenRef<Data>{ inner: inner } = ref;
+        TokenRef<BaseToken> { inner }
+    }
+
+    public fun from_base_token<Data: store>(ref: TokenRef<BaseToken>): TokenRef<Data> {
+        let addr = token_addr_from_ref(&ref);
+        assert!(exists<Token<Data>>(addr), EINVALID_REF_CONVERSION);
+        let TokenRef<BaseToken>{ inner: inner } = ref;
+        TokenRef<Data> { inner }
+    }
+
+    public fun token_addr_from_ref<Data>(token_ref: &TokenRef<Data>): address {
+        account::get_signer_capability_address(&token_ref.inner)
+    }
+
+    public fun token_signer<Data: store>(
+        creator: &signer,
+        token_ref: &TokenRef<Data>,
+    ): signer acquires Token {
+        let token = borrow_global<Token<Data>>(token_addr_from_ref(token_ref));
+        assert!(token.creator == signer::address_of(creator), ENOT_THE_CREATOR);
+        account::create_signer_with_capability(&token_ref.inner)
+    }
+
+    public fun set_data<Data: store>(ref: &TokenRef<Data>, data: Data) acquires Token {
+        let token = borrow_global_mut<Token<Data>>(token_addr_from_ref(ref));
+        option::fill(&mut token.data, data)
+    }
+
+    public fun take_data<Data: store>(ref: &TokenRef<Data>): Data acquires Token {
```

Comment:
Biggest restriction is that we can't have `borrow_data` or `borrow_data_mut`.

Reply:
:100: , but that's something that if the language can improve upon, we can backdoor through an upgrade :D